### PR TITLE
mgr/cephadm: move NodeAssignment to new python module

### PR DESCRIPTION
**Pull request by sebastian-philipp**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by sebastian-philipp** [comment]
```
2020-02-03T16:02:45.453+0100 7f9039bf1700 -1 Remote method threw exception: Traceback (most recent call last):
  File "/home/user/src/ceph/src/pybind/mgr/cephadm/module.py", line 1839, in add_mds
    raise orchestrator.OrchestratorValidationError("must specify at least %d hosts" % count)
orchestrator.OrchestratorValidationError: must specify at least 0 hosts

2020-02-03T16:02:45.453+0100 7f9039bf1700  0 [volumes] [ERROR] [volumes.fs.fs_util] Failed to create MDS daemons
Traceback (most recent call last):
  File "/home/user/src/ceph/src/pybind/mgr/volumes/fs/fs_util.py", line 39, in create_mds
    completion = mgr.add_mds(spec)
  File "/home/user/src/ceph/src/pybind/mgr/orchestrator.py", line 1445, in inner
    completion = self._oremote(method_name, args, kwargs)
  File "/home/user/src/ceph/src/pybind/mgr/orchestrator.py", line 1515, in _oremote
    return mgr.remote(o, meth, *args, **kwargs)
  File "/home/user/src/ceph/src/pybind/mgr/mgr_module.py", line 1429, in remote
    args, kwargs)
RuntimeError: Remote method threw exception: Traceback (most recent call last):
  File "/home/user/src/ceph/src/pybind/mgr/cephadm/module.py", line 1839, in add_mds
    raise orchestrator.OrchestratorValidationError("must specify at least %d hosts" % count)
orchestrator.OrchestratorValidationError: must specify at least 0 hosts

2020-02-03T16:02:45.453+0100 7f9039bf1700 -1 mgr.server reply reply (22) Invalid argument Remote method threw exception: Traceback (most recent call last):
  File "/home/user/src/ceph/src/pybind/mgr/cephadm/module.py", line 1839, in add_mds
    raise orchestrator.OrchestratorValidationError("must specify at least %d hosts" % count)
orchestrator.OrchestratorValidationError: must specify at least 0 hosts
```

- [ ] TODO: `s/0/1/`

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1693,8 +1694,9 @@ def update_mons(self, spec):
             # Improve Error message. Point to parse_host_spec examples
             raise orchestrator.OrchestratorValidationError("Mons need a host spec. (host, network, name(opt))")
 
-        spec = NodeAssignment(spec=spec, get_hosts_func=self._get_hosts, service_type='mon').load()
-        return self._update_mons(spec)
+        assigned_spec = NodeAssignment(spec=spec, get_hosts_func=self._get_hosts, service_type='mon').load()
+        assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
```
Comment (liewegas) [inline src/pybind/mgr/cephadm/module.py]:
```
2020-02-04T17:49:58.192 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: debug 2020-02-04T17:49:58.191+0000 7f74e470a700 -1 Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.192 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: AssertionError
2020-02-04T17:49:58.193 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: debug 2020-02-04T17:49:58.191+0000 7f74e470a700 -1 mgr handle_command module 'orchestrator_cli' command handler threw exception: Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.194 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.194 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.195 INFO:ceph.mgr.y.smithi162.stdout:Feb 04 17:49:58 smithi162 bash[5452]: AssertionError
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:Error EINVAL: Traceback (most recent call last):
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1069, in _handle_command
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:    return CLICommand.COMMANDS[cmd['prefix']].call(self, cmd, inbuf)
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 309, in call
2020-02-04T17:49:58.195 INFO:teuthology.orchestra.run.smithi162.stderr:    return self.func(mgr, **kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 140, in wrapper
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    return func(*args, **kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator_cli/module.py", line 641, in _update_mons
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    completion = self.update_mons(spec)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 1445, in inner
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    completion = self._oremote(method_name, args, kwargs)
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/orchestrator.py", line 1515, in _oremote
2020-02-04T17:49:58.196 INFO:teuthology.orchestra.run.smithi162.stderr:    return mgr.remote(o, meth, *args, **kwargs)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1429, in remote
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:    args, kwargs)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:RuntimeError: Remote method threw exception: Traceback (most recent call last):
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:  File "/usr/share/ceph/mgr/cephadm/module.py", line 1700, in update_mons
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:    assert isinstance(assigned_spec, orchestrator.StatefulServiceSpec)
2020-02-04T17:49:58.197 INFO:teuthology.orchestra.run.smithi162.stderr:AssertionError
```
/a/sage-2020-02-04_16:34:01-rados-wip-sage-testing-2020-02-04-0730-distro-basic-smithi/4732250

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py]: Yay! this looks like a real thing! I always assumed that we're getting a StatefulServiceSpec

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/module.py]: #33175

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -1174,7 +1176,7 @@ def _run_cephadm(self, host, entity, command, args,
             self.log.exception(ex)
             raise
 
-    def _get_hosts(self, wanted=None):
+    def _get_hosts(self, wanted=None) -> List[str]:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py]: this is wrong

**Comment by sebastian-philipp** [comment]
jenkins test make check

**Comment by jschmid1** [comment]
```
cephadm/module.py: note: In member "add_prometheus" of class "CephadmOrchestrator":
cephadm/module.py:2174: error: Unexpected keyword argument "service_type" for "NodeAssignment"
cephadm/scheduling.py:60: note: "NodeAssignment" defined here
cephadm/module.py: note: In member "apply_prometheus" of class "CephadmOrchestrator":
cephadm/module.py:2204: error: Unexpected keyword argument "service_type" for "NodeAssignment"
cephadm/scheduling.py:60: note: "NodeAssignment" defined here
```

besides that, +1 for moving it into its separate module

**Comment by liewegas** [comment]
```
2020-02-20T22:35:38.547 INFO:teuthology.orchestra.run.smithi060.stderr:Error ENOENT: must specify at least 2 hosts
```
/a/sage-2020-02-20_22:15:03-rados-wip-sage-testing-2020-02-20-1323-distro-basic-smithi/4785144

**Comment by p-se** [comment]
jenkins test make check

**Comment by sebastian-philipp** [comment]
rebased

**Comment by liewegas** [comment]
- diff version of mon fix is part of #33496 
- diff version of schedule fix is part of #33465

**Comment by liewegas** [comment]
conflicts with #33523

**Comment by sebastian-philipp** [comment]
> conflicts with #33523

can't get in through QA due to missing containers.

**Comment by sebastian-philipp** [comment]
Frustrating. 

```
2020-02-28T14:15:38.740 INFO:teuthology.orchestra.run.smithi064.stderr:Error EINVAL: Traceback (most recent call last):
2020-02-28T14:15:38.740 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1133, in _handle_command
2020-02-28T14:15:38.740 INFO:teuthology.orchestra.run.smithi064.stderr:    return self.handle_command(inbuf, cmd)
2020-02-28T14:15:38.741 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 191, in handle_command
2020-02-28T14:15:38.741 INFO:teuthology.orchestra.run.smithi064.stderr:    return dispatch[cmd['prefix']].call(self, cmd, inbuf)
2020-02-28T14:15:38.741 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 308, in call
2020-02-28T14:15:38.741 INFO:teuthology.orchestra.run.smithi064.stderr:    return self.func(mgr, **kwargs)
2020-02-28T14:15:38.742 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 153, in <lambda>
2020-02-28T14:15:38.742 INFO:teuthology.orchestra.run.smithi064.stderr:    wrapper_copy = lambda *l_args, **l_kwargs: wrapper(*l_args, **l_kwargs)
2020-02-28T14:15:38.742 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 144, in wrapper
2020-02-28T14:15:38.742 INFO:teuthology.orchestra.run.smithi064.stderr:    return func(*args, **kwargs)
2020-02-28T14:15:38.742 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/module.py", line 734, in _apply_mgr
2020-02-28T14:15:38.743 INFO:teuthology.orchestra.run.smithi064.stderr:    completion = self.apply_mgr(spec)
2020-02-28T14:15:38.743 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 1757, in inner
2020-02-28T14:15:38.743 INFO:teuthology.orchestra.run.smithi064.stderr:    completion = self._oremote(method_name, args, kwargs)
2020-02-28T14:15:38.743 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/orchestrator/_interface.py", line 1827, in _oremote
2020-02-28T14:15:38.744 INFO:teuthology.orchestra.run.smithi064.stderr:    return mgr.remote(o, meth, *args, **kwargs)
2020-02-28T14:15:38.744 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/mgr_module.py", line 1495, in remote
2020-02-28T14:15:38.744 INFO:teuthology.orchestra.run.smithi064.stderr:    args, kwargs)
2020-02-28T14:15:38.744 INFO:teuthology.orchestra.run.smithi064.stderr:RuntimeError: Remote method threw exception: Traceback (most recent call last):
2020-02-28T14:15:38.745 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/cephadm/module.py", line 2154, in apply_mgr
2020-02-28T14:15:38.745 INFO:teuthology.orchestra.run.smithi064.stderr:    return self._apply_service('mgr', spec, self._create_mgr)
2020-02-28T14:15:38.745 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/cephadm/module.py", line 2010, in _apply_service
2020-02-28T14:15:38.745 INFO:teuthology.orchestra.run.smithi064.stderr:    service_type=daemon_type).load()
2020-02-28T14:15:38.746 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/cephadm/scheduling.py", line 78, in load
2020-02-28T14:15:38.746 INFO:teuthology.orchestra.run.smithi064.stderr:    self.validate_placement()
2020-02-28T14:15:38.746 INFO:teuthology.orchestra.run.smithi064.stderr:  File "/usr/share/ceph/mgr/cephadm/scheduling.py", line 139, in validate_placement
2020-02-28T14:15:38.746 INFO:teuthology.orchestra.run.smithi064.stderr:    raise orchestrator.OrchestratorValidationError("must specify at least %d hosts" % count)
2020-02-28T14:15:38.747 INFO:teuthology.orchestra.run.smithi064.stderr:orchestrator._interface.OrchestratorValidationError: must specify at least 2 hosts
```

**Pull request closed without merging**